### PR TITLE
pnpm.fetchDeps: Add workspace and custom pnpm config commands support

### PR DESCRIPTION
**Pull request by pyrox0**
## Description of changes
Adds `pnpmWorkspace` and `extraPnpmCommands` options for pnpm.fetchDeps calls

These can be used to either use a specific pnpm workspace for installation, and to run custom pnpm config commands during dependency fetching. Some packages may need a setting such as `dedupe-peer-dependents` set to a non-default value, and so this allows them to set those when needed.

See the discussion in #309100 (starting at https://github.com/NixOS/nixpkgs/pull/309100#discussion_r1658477855) as to why these are useful.

I've also updated the documentation appropriately.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: doc/languages-frameworks/javascript.section.md
Comment (doronbehar) [inline doc/languages-frameworks/javascript.section.md:376]:
```suggestion
#### PNPM Workspaces {#javascript-pnpm-workspaces}
```

File: doc/languages-frameworks/javascript.section.md
Comment (doronbehar) [inline doc/languages-frameworks/javascript.section.md:393]:
```suggestion
#### Additional PNPM Commands {#javascript-pnpm-extraCommands}
```

File: doc/languages-frameworks/javascript.section.md
Comment (doronbehar) [inline doc/languages-frameworks/javascript.section.md:400]: I'd name this `prePnpmInstall`, to make it similar to the `preInstall` hook, which can't be used in this example.

Another option, which can be implemented in addition is an interface like this:

```nix
# ...
pnpmDeps = pnpm.fetchDeps {
  # ...
  extraPnpmConfig = {
    dedupe-peer-dependants = "false";
  };
};
```

This should be easily implemented via something like this (not tested):

```nix
let 
  prePnpmInstall = lib.pipe [
    (lib.mapAttrsToList (n:v: "pnpm config set ${n} ${builtins.toString v}"))
    (lib.concatStringsSep "\n")
  ] extraConfig;
in
  stdenvNoCC.mkDerivation # ...
  # ...
    installPhase = ''
      # ...
    '';
  # ...
```

You could also combine both interfaces by allowing both a string `prePnpmInstall` and a declerative `extraPnpmConfig` attribute set.

**Comment by doronbehar** [comment]
On a 2nd thought, this PR is a good opportunity to test this new interface. I also had motivation to work on this feature for `bash-language-server` and to fix #316908 , so I pushed 2 more commits that will also close #309100 .

File: doc/languages-frameworks/javascript.section.md
Comment (doronbehar) [reply doc/languages-frameworks/javascript.section.md]: > Another option, which can be implemented in addition is an interface like this:

After playing with this interface a bit I reached the conclusion that it will be a bit confusing to have one declarative argument that only fits the `pnpm.fetchDeps` arguments, and that can't be used as an attribute set argument to the `mkDerivation` builder - because it is an attribute set.

Hence I renamed the string option to `prePnpmInstall` and I'm marking this conversation as resolved.

File: pkgs/development/tools/pnpm/fetch-deps/default.nix
Comment (getchoo) [inline pkgs/development/tools/pnpm/fetch-deps/default.nix:36]:
```suggestion
      installFlags = lib.optional (pnpmWorkspace != "")  "--filter=${pnpmWorkspace}";
```
this would also be nice as similar to above, callers would be able to use standard `mkDerivation` variables rather than unique ones found here

File: pkgs/development/tools/pnpm/fetch-deps/default.nix
Comment (getchoo) [inline pkgs/development/tools/pnpm/fetch-deps/default.nix:67]:
```suggestion
```
see above. this wouldn't be required anymore

File: pkgs/development/tools/pnpm/fetch-deps/default.nix
Comment (getchoo) [inline pkgs/development/tools/pnpm/fetch-deps/default.nix:70]:
```suggestion
          pnpm install --frozen-lockfile --ignore-script --force "$installFlags"
```

File: pkgs/development/tools/pnpm/fetch-deps/pnpm-config-hook.sh
Comment (getchoo) [inline pkgs/development/tools/pnpm/fetch-deps/pnpm-config-hook.sh:28]:
```suggestion
        pnpmInstallFlags+=("--filter=$pnpmWorkspace")
```

File: pkgs/development/tools/pnpm/fetch-deps/pnpm-config-hook.sh
Comment (getchoo) [inline pkgs/development/tools/pnpm/fetch-deps/pnpm-config-hook.sh:32]:
```suggestion
    pnpm install --offline --frozen-lockfile --ignore-script "$pnpmInstallFlags"
```

**Review by getchoo — CHANGES_REQUESTED** [review]

File: pkgs/development/tools/pnpm/fetch-deps/pnpm-config-hook.sh
Comment (doronbehar) [reply pkgs/development/tools/pnpm/fetch-deps/pnpm-config-hook.sh:28]: Nice, thanks!

File: pkgs/development/tools/pnpm/fetch-deps/default.nix
Comment (doronbehar) [reply pkgs/development/tools/pnpm/fetch-deps/default.nix:36]: > this would also be nice as similar to above, this would also be nice as similar to above, callers would be able to use standard `mkDerivation` variables rather than unique ones found here

I agree that `installFlags` is a better variable name (pushed this change), but I'm not sure how that helps someone override this variable with an interface similar to `mkDerivation`...

File: pkgs/development/tools/pnpm/fetch-deps/default.nix
Comment (doronbehar) [reply pkgs/development/tools/pnpm/fetch-deps/default.nix:67]: Another option would be to `runHook prePnpmInstall` here.

File: pkgs/development/tools/pnpm/fetch-deps/pnpm-config-hook.sh
Comment (doronbehar) [reply pkgs/development/tools/pnpm/fetch-deps/pnpm-config-hook.sh:32]: Nice, this also requires to document this variable and to allow others to use it. I revised the first commit to include this change, and I tested and documented this interface.

File: pkgs/development/tools/pnpm/fetch-deps/default.nix
Comment (doronbehar) [reply pkgs/development/tools/pnpm/fetch-deps/default.nix:18]: > i think it would be much better to just use a built in phase here
> 
> i.e., if someone wants to set specific options or run arbitrary commands, they can just use `postConfigure`/`postInstall`/etc from `mkDerivation` rather than this (single) unique phase
> 
> this would require splitting up some of the internal phases here, namely when `HOME`, `store-dir`, and other defaults are applied - they should probably in the `configurePhase`

Hmm you have a point, but still though, even if I'd simply move the `runHook preInstall` line after `HOME` is set, which would allow to perform such an override, I'd have to do something like this:

```nix
stdenv.mkDerivation (finalAttrs: {
  # ...
  pnpmDeps = (pnpm_8.fetchDeps {
    inherit (finalAttrs) pname version src pnpmWorkspace;
    hash = "sha256-n7HTd/rKxJdQKnty5TeOcyvBU9j/EClQ9IHqbBaEwQE=";
  }).overrideAttrs(_: {
    preInstall = finalAttrs.prePnpmInstall;
  });

  prePnpmInstall = ''
    pnpm config set dedupe-peer-dependents false
  '';
  # ...
})
```

Seems much more ugly. Even if we'll just rename the argument to `preInstall`, such that the interface would look like this:

```nix
stdenv.mkDerivation (finalAttrs: {
  # ...
  pnpmDeps = pnpm_8.fetchDeps {
    inherit (finalAttrs) pname version src pnpmWorkspace;
    preInstall = finalAttrs.prePnpmInstall;
    hash = "sha256-n7HTd/rKxJdQKnty5TeOcyvBU9j/EClQ9IHqbBaEwQE=";
  };

  prePnpmInstall = ''
    pnpm config set dedupe-peer-dependents false
  '';
  # ...
})
```

It is still a bit weird IMO how these arguments are named differently, since they serve exactly the same purpose.

If you weren't thinking of either of these interfaces, I didn't fully understand you.

File: doc/languages-frameworks/javascript.section.md
Comment (SuperSandro2000) [inline doc/languages-frameworks/javascript.section.md:395]: Why can't we detect pnpmWorkspace and set that automatically in the hook?

File: pkgs/development/tools/pnpm/fetch-deps/pnpm-config-hook.sh
Comment (SuperSandro2000) [inline pkgs/development/tools/pnpm/fetch-deps/pnpm-config-hook.sh:29]: If we set this already, why are we doing the buildPhases above by hand? I am confused

**Comment by dezren39** [comment]
@doronbehar would this be something we could use for wrangler in #322573 ? should we wait for this to possibly merge in and then update the pr with these new commands? the pr builds and works fine, but uses quite a bit of disk and needs a bit more testing.

File: doc/languages-frameworks/javascript.section.md
Comment (dezren39) [inline doc/languages-frameworks/javascript.section.md:395]:
```suggestion
Usually in such cases, you'd want to use `pnpm --filter=$pnpmWorkspace build` to build your project, as `npmHooks.npmBuildHook` probably won't work. A `buildPhase` based on the following example will probably fit most workspace projects:
```

File: pkgs/development/tools/pnpm/fetch-deps/pnpm-config-hook.sh
Comment (doronbehar) [reply pkgs/development/tools/pnpm/fetch-deps/pnpm-config-hook.sh:29]: > If we set this already, why are we doing the buildPhases above by hand? I am confused

You are confused because (surprisingly) we don't have a `pnpm.buildHook`. If the project you try to package doesn't have a specific workspace you are interested in, using `npmHooks.npmBuildHook` probably will do the job. With a specific workspace you pretty much have to use `pnpm --filter=.. build`. Perhaps we should add `pnpm.buildHook`.

File: doc/languages-frameworks/javascript.section.md
Comment (doronbehar) [reply doc/languages-frameworks/javascript.section.md:395]: > Why can't we detect pnpmWorkspace and set that automatically in the hook?

See my comment regarding the `pnpm.buildHook` idea below.

**Comment by thenbe** [comment]
Just a heads up that pnpm had a [bug](https://github.com/pnpm/pnpm/issues/6300) where attempting to install deps using `--filter` wasn't behaving as expected in certain cases.

The issue was fixed in pnpm `v9.5.0`. nixos-unstable is currently on `v9.4.0` (an update [PR](https://github.com/NixOS/nixpkgs/pull/325837) to `v9.6.0` is pending).

**Comment by doronbehar** [comment]
Thanks @thenbe for notifying. Now (after https://github.com/NixOS/nixpkgs/pull/325837 is merged) this is rebased. I will merge this after CI is green.

**Comment by doronbehar** [comment]
There are a few unresolved review conversations above, so I let @SuperSandro2000 and @getchoo a chance to reply to them for a few days.

File: doc/languages-frameworks/javascript.section.md
Comment (gepbird) [inline doc/languages-frameworks/javascript.section.md:395]:
```suggestion
Note that you do not need to set `sourceRoot` to make this work.
```

File: doc/languages-frameworks/javascript.section.md
Comment (gepbird) [inline doc/languages-frameworks/javascript.section.md:424]:
```suggestion
In this example, `prePnpmInstall` will be run by both `pnpm.configHook` and by the `pnpm.fetchDeps` builder.
```

**Review by gepbird — CHANGES_REQUESTED** [review]

**Comment by doronbehar** [comment]
Thanks for the suggestions @gepbird! Both of them fixed.

File: pkgs/development/tools/pnpm/fetch-deps/default.nix
Comment (niklaskorz) [inline pkgs/development/tools/pnpm/fetch-deps/default.nix:68]: I know this is technically unrelated to this PR, but there is a typo here, the parameter is called `--ignore-scripts`: https://pnpm.io/cli/install#--ignore-scripts

File: pkgs/development/tools/pnpm/fetch-deps/pnpm-config-hook.sh
Comment (niklaskorz) [inline pkgs/development/tools/pnpm/fetch-deps/pnpm-config-hook.sh:32]: Same here (`--ignore-scripts`)

**Comment by doronbehar** [comment]
Thanks @niklaskorz for the review :) I fixed that in separate commits before the main commits of the PR. I will merge this when CI is green.

**Comment by ck3d** [comment]
This PR broke the package bash-language-server, which has been fixed with #333701.

**Comment by ck3d** [comment]
The resulting dependencies for the packages `bash-language-server` and `astro-language-server` do not change if `pnpmWorkspace` is set. The hashes and the content of the dependencies is the same, regardless of the value of `pnpmWorkspace`.
I think this PR needs rework for the parameter `pnpmWorkspace`.

**Comment by doronbehar** [comment]
> The resulting dependencies for the packages `bash-language-server` and `astro-language-server` do not change if `pnpmWorkspace` is set. The hashes and the content of the dependencies is the same, regardless of the value of `pnpmWorkspace`. I think this PR needs rework for the parameter `pnpmWorkspace`.

Thanks for noticing that. I suspect that what me and @pyrox0 experienced back when we initially worked on this PR, was a related to [this upstream pnpm issue](https://github.com/pnpm/pnpm/issues/6300). Perhaps indeed now the workspace support could be simplified. I'm pretty sure though, that for `astro-language-server`, using `pnpm config set dedupe-peer-dependants false` is required, and support for `prePnpmInstall` was added here as well.

Hence for now, I think we'll have to wait for more packages that use pnpm workspaces to aggregate in Nixpkgs, and we can revise this topic.

**Pull request merged**
